src/net/Connection.cpp

```diff
@@ -0,0 +1,49 @@
+#include "Connection.h"
+
+tcp::socket& Connection::get_socket() {
+    return socket;
+}
+
+void Connection::start_async_read_loop() {
+    socket.async_read_some(boost::asio::buffer(rcvbuf),
+        [&](const boost::system::error_code& error, size_t n_bytes) {
+        if (!error) {
+            message_queue.push(rcvbuf);
+        } else if (error != boost::asio::error::eof) {
+            std::cerr << "FATAL handle_read error: " << error << "\n";
+            return;
+        }
+
+        // Wait for and read the next message.
+        start_async_read_loop();
+    });
+}
+
+bool Connection::send(const string& message) {
+    if (!socket.is_open()) {
+        return false;
+    }
+
+    boost::system::error_code error;
+    boost::asio::write(socket, boost::asio::buffer(message), error);
```

Comment:
Since we're doing async read, might as well do async write with lambda? http://www.boost.org/doc/libs/1_41_0/doc/html/boost_asio/reference/async_write/overload1.html